src/electionguard/discrete_log.py

```diff
@@ -57,30 +84,67 @@ async def discrete_log_async(
     async with mutex:
         if element in cache:
             return (cache[element], cache)
+        if not lazy_evaluation:
+            raise DiscreteLogNotFoundError(element)
 
-        _cache = compute_discrete_log_cache(element, cache)
+        _cache = compute_discrete_log_cache(element, cache, max_exponent)
         return (_cache[element], _cache)
 
 
+def precompute_discrete_log_cache(
+    max_exponent: int, cache: DiscreteLogCache = None
+) -> DiscreteLogCache:
+    """
+    Precompute the discrete log by the max exponent.
+    """
+
+    if max_exponent > _DLOG_MAX_EXPONENT:
+        raise DiscreteLogExponentError(max_exponent)
+
+    if not cache:
+        cache = _INITIAL_CACHE
+
+    current_element = list(cache)[-1]
+    prev_exponent = cache[current_element]
+
+    if prev_exponent >= max_exponent:
+        return cache
+
+    g = ElementModP(get_generator(), False)
+
+    for exponent in range(prev_exponent + 1, max_exponent + 1):
+        current_element = mult_p(g, current_element)
+        cache[current_element] = exponent
+
+    return cache
+
+
 def compute_discrete_log_cache(
-    element: ElementModP, cache: DLogCache
-) -> Dict[ElementModP, int]:
+    element: ElementModP,
+    cache: DiscreteLogCache,
+    max_exponent: int = _DLOG_MAX_EXPONENT,
+) -> DiscreteLogCache:
     """
-    Compute a discrete log cache up to the specified element.
+    Compute or lazy evaluation a discrete log cache up to the specified element.
     """
+
+    if max_exponent > _DLOG_MAX_EXPONENT:
+        raise DiscreteLogExponentError(max_exponent)
+
     if not cache:
-        cache = {ONE_MOD_P: 0}
+        cache = _INITIAL_CACHE
+
     max_element = list(cache)[-1]
     exponent = cache[max_element]
 
     g = ElementModP(get_generator(), False)
+
     while element != max_element:
         exponent = exponent + 1
-        if exponent > DLOG_MAX:
-            raise ValueError("size is larger than max.")
+        if exponent > max_exponent:
+            raise DiscreteLogExponentError(exponent, max_exponent)
```

Comment:
Might be good to move this check right after line 138 to check the values before making the ElementModP.

Reply:
It adds a little bit of redundancy because we still need it here. Let's discuss.